### ISSUE TITLE
Unclear behavior of `then` inside pattern-matching `switch`

### ISSUE THREAD

**Issue by imm** [DanielXMoore/Civet#607]
TL;DR: It is not always possible to use `then` without `when` in `switch`, and when it is possible, the expression is not returned.

Let's take an example from the site:
```coffee
switch x
  % 15 is 0
    console.log "fizzbuzz"
  % 3 is 0
    console.log "fizz"
  % 5 is 0
    console.log "buzz"
  else
    console.log x
```

Now I want to make the code more compact
```coffee
switch x
  % 15 is 0 then console.log "fizzbuzz"
  % 3 is 0 then console.log "fizz"
               ^ ParseError
```

Okay, so you can't do that, which is very disappointing. With the keyword `when` it was possible, but without pattern-matching capabilities. The compiler, for some reason, indicates an error in the second line, and not in the first.

Probably for the same reason, in the last condition you can use `then`:
```coffee
switch x
  % 15 is 0
    console.log "fizzbuzz"
  % 3 is 0
    console.log "fizz"
  % 5 is 0 then console.log "buzz"
  else
    console.log x
```
Compiles as expected. It's confusing.

But that is not all. Let's say I decide to make the code even more compact (why repeat `console.log` so many times?):
```coffee
console.log switch x
  % 15 is 0
    "fizzbuzz"
  % 3 is 0
    "fizz"
  % 5 is 0
    "buzz"
  else x
```
I still can't use `then` everywhere except for the last condition, but if I do that, I get unexpected results.
```coffee
console.log switch x
  % 15 is 0
    "fizzbuzz"
  % 3 is 0
    "fizz"
  % 5 is 0 then "buzz"
  else x
```
compiles to
```js
console.log(
  (() => {
    if (x % 15 === 0) {
      return "fizzbuzz";
    } else if (x % 3 === 0) {
      return "fizz";
    } else if (x % 5 === 0) {
      ("buzz"); // <- `return` is missing
    } else {
      return x;
    }
  })()
);
```
The last condition stopped returning the value. It would be preferable that either `then` work in all conditions or none. And if the use of `then` would be allowed, then the result would still be returned. In any case, it would be cool to be able to write the condition and the result on the same line.

The issue is based on real events. In my case it was a function that was supposed to return an array. So I used `for` with nested `switch`. The function turned out not very big, but it took a lot of lines. Unfortunately, only the last condition could be reduced to one line, and this simple action turned the proven working code into the wrong one (`results.push` disappeared). The reasons were very obscure. Sorry for this psychotherapeutic part.

P.S. Here is another example of a valid use of `then`:
```coffee
console.log switch x
  % 15 is 0 then ;
  % 13 is 0 then ->
  % 7 is 0 then do
    nonReturnableValue
  % 7 is 0
    do
      returnableValue
  % 3 is 0 # but not here
    "fizz"
  % 5 is 0 then "buzz"
  else x
```

**Comment by edemaine** [comment]
Thanks for reporting! That's definitely bad behavior.

One view of the issue here is that we shouldn't allow binary-op next-line continuations in a single-line statement (the body of a `then`). You can see how this is causing problems in [this example](https://civet.dev/playground/?code=c3dpdGNoIHgKICAlIDE1IGlzIDAgdGhlbiBjb25zb2xlLmxvZyAiZml6emJ1enoiCiAgJSAzIGlzIDA%3D):

```js
switch x
  % 15 is 0 then console.log "fizzbuzz"
  % 3 is 0
↓↓↓
if (x % 15 === 0) {
  console.log("fizzbuzz" % 3 === 0);
}
```

An alternative view is that `then` should require indentation of at least +1 beyond the first line. But I think it's really just intended for single-line content.
